### ISSUE TITLE
Disabling email and phone login

### ISSUE THREAD

**Issue by ideen1** [aws-amplify/amplify-backend#1313]
### Environment information

```plain text
System:
  OS: Windows 10 10.0.19045
  CPU: (16) x64 11th Gen Intel(R) Core(TM) i9-11900K @ 3.50GHz
  Memory: 48.14 GB / 63.81 GB
Binaries:
  Node: 20.9.0 - C:\Program Files\nodejs\node.EXE
  Yarn: undefined - undefined
  npm: 10.1.0 - C:\Program Files\nodejs\npm.CMD
  pnpm: undefined - undefined
NPM Packages:
  @aws-amplify/backend: 0.13.0-beta.6
  @aws-amplify/backend-cli: 0.12.0-beta.6
  aws-amplify: 6.0.18
  aws-cdk: Not Found
  aws-cdk-lib: 2.131.0
  typescript: 5.3.3
AWS environment variables:
  AWS_NODEJS_CONNECTION_REUSE_ENABLED = 1
  AWS_SDK_LOAD_CONFIG = 1
  AWS_STS_REGIONAL_ENDPOINTS = regional
No CDK environment variables
```

### Description

When setting the `loginWith` properties of `defineAuth`, we should be able to disable both email and phone if `externalProviders` are set. Currently, we get a warning saying either phone or email login must be enabled. 

We can disable them manually from the console. However, subsequent updates will override that console settings so it would be ideal to have them disabled in our `defineAuth` code.

**Comment by josefaidt** [comment]
Hey @ideen1 :wave: thanks for raising this! Are you looking to only allow signing in via an external provider? Unfortunately these values cannot be changed after the initial deployment, however you can disable user sign-ups, or use a pre-sign-up trigger to customize the desired sign-up experience

```ts
// amplify/backend.ts
import { defineBackend } from "@aws-amplify/backend"
import { auth } from "./auth/resource"
import { data } from "./data/resource"

const backend = defineBackend({
  auth,
  data,
})

const { cfnUserPool } = backend.auth.resources.cfnResources

cfnUserPool.adminCreateUserConfig = {
  allowAdminCreateUserOnly: true,
}
```

**Comment by ideen1** [comment]
Yeah, certain enterprise applications require only externalProviders to be enabled. If it’s not something that can be changed after the first deployment, is it something that can be set before the first deployment at least? I don’t imagine needing to turn that back on if that’s what the user wants to do initially. I have already disabled user sign ups as you showed and that’s is a great step.

**Comment by josefaidt** [comment]
Hey @ideen1 thanks for clarifying! That makes total sense 🙂 the Cognito User Pool's [username attributes](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html#user-pool-settings-aliases) cannot be changed after deployment, where this corresponds to `loginWith.email` and `loginWith.phone` in `defineAuth`. Disabling user self sign-up will block attempts at signing up, but allows user creation with the [AdminCreateUser API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminCreateUser.html) in the event you'd like to create users manually. (this permission can also be denied for your AWS users)

An alternative is to use a custom Identity Pool to allow users to federate but with the caveat that a Cognito-native user will not be created. The Cognito-native users will allow you to bind specific attributes from your upstream identity provider to the Cognito user, bucket users into groups, etc.

**Comment by ideen1** [comment]
Right I see, perhaps what I'm asking is to disable it on the Hosted UI. Then, the login form does not appear on the Hosted UI. I know this is possible through the console on the Hosted UI settings page even after the UserPool as been deployed. Is this something we could do from our Amplify backend configuration?

**Comment by ideen1** [comment]
> Hey @ideen1 👋 thanks for raising this! Are you looking to only allow signing in via an external provider? Unfortunately these values cannot be changed after the initial deployment, however you can disable user sign-ups, or use a pre-sign-up trigger to customize the desired sign-up experience
> 
> ```ts
> // amplify/backend.ts
> import { defineBackend } from "@aws-amplify/backend"
> import { auth } from "./auth/resource"
> import { data } from "./data/resource"
> 
> const backend = defineBackend({
>   auth,
>   data,
> })
> 
> const { cfnUserPool } = backend.auth.resources.cfnResources
> 
> cfnUserPool.adminCreateUserConfig = {
>   allowAdminCreateUserOnly: true,
> }
> ```

@josefaidt Also, just to confirm, the `allowAdminCreateUserOnly` field is marked as readonly as per the type definition, so will this actually make the change we expect? Maybe the readonly is a mistake?

**Comment by ykethan** [comment]
Hey the Gen 2 documentation has been updated, do checkout at https://docs.amplify.aws/react/build-a-backend/auth/concepts/usernames/ and let us know this assists in building your use case.

**Comment by ideen1** [comment]
@ykethan Correct me if I am wrong but I believe that link is more related to usernames. My original question was the ability to completely remove email/password sign on from the UserPool during the initial deployment. @josefaidt Mentioned that it can not be done after initial deployment but I'm wondering if it can be done during the initial deployment.

The thing that is missing here is fully enforcing that no user can ever be able to "self-sign-up" .As mentioned above, the `allowAdminCreateUserOnly` is marked readonly, so I am not sure if this is safe to use.  It may be ideal to have something as crucial as this added as a argument to defineAuth.

**Comment by josefaidt** [comment]
Hey @ideen1 :wave: yes, while that property of `adminCreateUserConfig` is readonly, the parent is not and we can safely override it with our own settings. Using the snippet above, when sandbox is running we can inspect the rendered CloudFormation template in the artifacts at `.amplify/artifacts`
![CleanShot 2024-05-08 at 10 11 03](https://github.com/aws-amplify/amplify-backend/assets/5033303/096dcf55-62ac-4ccb-9413-a5c42c7f0b09)

**Comment by ykethan** [comment]
Closing the issue due to inactivity. Do reach out to us if you are still experiencing this issue.
